File: Homework1/CodesOld/22202709_Ozongun_HalilArda.go
package main

import "fmt"
import "reflect"
import "sort"

/*
What types are legal for subscripts?
Are subscripting expressions in element references range checked?
Are ragged multidimensional arrays allowed?
Can array objects be initialized?
Are any kind of slices supported?
Which operators are provided?
*/


func main() {

	fmt.Println("1: Legal types for subscripts: ")

	arr1 := [5]int{1,2,3,4,5}
	fmt.Println(arr1[3])
	var index int64 = 4
	fmt.Println(arr1[index]) // 64 bit integer is legal
	fmt.Println(arr1[int64(3)]) 
	//fmt.Println(arr1[3.3]) -> Compile time Error: invalid array index 3.3 (type float64)
	fmt.Println(arr1[3.0]) // Legal: 3.0 is implicitly treated as int
	//fmt.Println(arr1["3"]) -> Compile time Error: invalid array index "3" (type string)
	var index3 float64 = 3.3
	fmt.Println(arr1[int(index3)]) // No error since float is converted to int
	//fmt.Println(arr1[index3])


	fmt.Println("2: Range checking in Subscripting expressions: ")
	// fmt.Println(arr1[9]) // Compile time error for constant index out of range
	//fmt.Println(arr1[-2]) 
	var index2 = 9
	fmt.Println(arr1[index2]) // Runtime error for variable index out of range


	
	fmt.Println("3: Ragged multidimensional arrays: ")
	
	var multidimensionalArr = [3][2]int{{1,2}, {3,4}, {6,7}} // Normal multidimensional array
	//var multidimensionalArrOOB = [3][2]int{{1,2}, {3,4}, {6,7,8}} // Compile time error.
	fmt.Println(multidimensionalArr)

	fmt.Println(reflect.TypeOf(multidimensionalArr).Kind())
	fmt.Println(reflect.TypeOf(multidimensionalArr[0]).Kind())

	var raggedArr = [3][4]int{{1,2,3,4}, {5,6}, {7,8,9}} // Legal. but not ragged. It fills with 0
	fmt.Println(raggedArr)
	fmt.Println(reflect.TypeOf(raggedArr).Kind())
	fmt.Println(reflect.TypeOf(raggedArr[1]).Kind())

	var raggedSlice = [3][]int{{1,2}, {3,4,5}, {6,7,8,9}} // If size is not given, it is array of slices
	fmt.Println(raggedSlice)
	fmt.Println(reflect.TypeOf(raggedSlice).Kind())
	fmt.Println(reflect.TypeOf(raggedSlice[1]).Kind())


	fmt.Println("4: Initialization of arrays: ")
	
	var arr2 = [5]int{1,2,3,4,5}
	fmt.Println(reflect.TypeOf(arr2).Kind())

	var arr3 = [...]int{1,2,3,4,5}
	fmt.Println(reflect.TypeOf(arr3).Kind())
	// Initialization with only size
	var arr4 = [5]int{}
	fmt.Println(arr4)
	fmt.Println(reflect.TypeOf(arr4).Kind())

	var arr5 [5]int = [5]int{1,2,3,4,5}
	fmt.Println(reflect.TypeOf(arr5).Kind())
	

	fmt.Println("5: Array Slice: ")

	slice1 := arr1[1:3] // Slice of array
	slice2 := arr1[1:] // Slice of array with capacity
	fmt.Println(slice1)
	fmt.Println(reflect.TypeOf(slice1).Kind())
	fmt.Println(slice2)
	fmt.Println(reflect.TypeOf(slice2).Kind())

	// step slicing not legal
	// slice3 := arr1[1:3:1] -> Compile time error.
	// slicing with function not legal

	fmt.Println("6: Allowed Operations: ")

	a1 := [3]int{1, 2, 3}
	a2 := [3]int{1, 2, 3}
	a3 := [3]int{4, 5, 6}

	// Equality
	fmt.Println(a1 == a2, a1 == a3) 

	// Arithmetic op.
	//fmt.Println(a1 + a2)
	//fmt.Println(a1 * a2)

	// Concatenation
	s1 := append(a1[:], a2[:]...)
	fmt.Println(s1)
	fmt.Println(reflect.TypeOf(a1[:]).Kind())
	fmt.Println(reflect.TypeOf(s1).Kind())
	
	// length 
	fmt.Println(len(a1))

	// To sort: 
	a1 = [3]int{3, 2, 1}
	fmt.Println(a1)
	sort.Ints(a1[:])
	fmt.Println("After sorting: ", a1)
	fmt.Println(reflect.TypeOf(a1).Kind())

	// Sum, min, max -> Not allowed.


}

File: Homework1/CodesOld/22202709_Ozongun_HalilArda.js
/*
1. What types are legal for subscripts?
2. Are subscripting expressions in element references range checked?
3. Are ragged multidimensional arrays allowed?
4. Can array objects be initialized?
5. Are any kind of slices supported?
6. Which operators are provided?
*/

console.log("1: Legal types for subscripts: ");

let arr1 = [1, 2, 3, 4, 5];
console.log(arr1[2]); // 3

let idx = 3;
console.log(arr1[idx]); // 4

// JS coerces strings to numbers if possible
console.log(arr1["1"]); // 2
console.log(arr1[1.0]); // 2

// Invalid types become undefined
console.log(arr1[true]); // undefined (true is coerced to "true")
console.log(arr1["banana"]); // undefined
console.log(arr1[[1]]); // 2 (array [1] coerced to "1")


console.log("\n2: Range checking in element references:");

console.log(arr1[10]); // undefined
console.log(arr1[-1]); // undefined (no negative indexing)


console.log("\n3: Ragged multidimensional arrays:");

let ragged = [
    [1, 2, 3],
    [4, 5],
    [6, 7, 8, 9]
];
console.log(ragged);

// Accessing elements
console.log(ragged[0][1]); // 2
console.log(ragged[1][2]); // undefined


console.log("\n4: Initialization of arrays:");

let init1 = [1, 2, 3];
let init2 = Array(5).fill(0);
let init3 = Array.from({length: 5}, (_, i) => i + 1);
let init4 = new Array(3); // undefined values

console.log(init1);
console.log(init2);
console.log(init3);
console.log(init4);


console.log("\n5: Array Slice:");

let sliced1 = arr1.slice(1, 3); // [2, 3]
let sliced2 = arr1.slice(2);    // [3, 4, 5]
let sliced3 = arr1.slice(-2);   // [4, 5]

console.log(sliced1);
console.log(sliced2);
console.log(sliced3);


console.log("\n6: Allowed Operations:");

let a1 = [1, 2, 3];
let a2 = [1, 2, 3];
let a3 = [4, 5, 6];

// Equality check
console.log(JSON.stringify(a1) === JSON.stringify(a2)); // true

// No built-in array arithmetic
// console.log(a1 + a3); // "1,2,34,5,6" (coerced to strings)

// Concatenation
let concat = a1.concat(a3);
console.log(concat);

// Length
console.log(a1.length); // 3

// Sorting
let unsorted = [3, 1, 2];
unsorted.sort();
console.log("Sorted:", unsorted);

// Sum, Min, Max
let sum = a1.reduce((acc, val) => acc + val, 0);
let max = Math.max(...a1);
let min = Math.min(...a1);

console.log("Sum:", sum);
console.log("Max:", max);
console.log("Min:", min);
